### PR TITLE
H3r01a/feature/rabbit

### PR DESCRIPTION
**Pull request by H3R01A**
Updated to branch to get asynchronous factory functionality to work!

### CODE REVIEW

File: packages/library/src/brokers/rabbit/main.ts [hunk 1]
```diff
@@ -70,40 +95,96 @@ export default class Rabbit extends MessageBroker {
...
+    //       that.topics[topic].exchange.name,
+    //       that.topics[topic].key ?? topic
+    //     );
+    //   });
+
+    console.log('in constructor - channel', that.channel);
+
+    console.log('in constructor - connection', that.connection);
+
+    console.log('this is that in the constructor', that);
```
Comment (mjhult) [inline packages/library/src/brokers/rabbit/main.ts:135]: If you don't mind removing zombie/debug code.

File: packages/library/src/brokers/rabbit/main.ts [hunk 2]
```diff
@@ -43,18 +43,43 @@ export type RabbitListenerOptions = GenericListenerOptions & {
...
 
-// TODO: Add error handling
+//function to act as asynchronous factory function
+/**
+ * Purpose: to produce new RabbitMQ message broker
+ * @param connection
+ * @param topics
+ * @returns RabbitMQ object to be used as a message broker
+ */
+export async function createRabbitClass(
```
Comment (mjhult) [inline packages/library/src/brokers/rabbit/main.ts:53]: We may want to rename this to just ```Rabbit```

**Review by mjhult — CHANGES_REQUESTED** [review]
Amazing! Thank you!

Just needs zombie/debug code removed

File: packages/library/src/brokers/rabbit/main.ts [hunk 3]
```diff
@@ -70,40 +95,96 @@ export default class Rabbit extends MessageBroker {
...
+    });
   }
 
   // TODO: Add support for multi messages
-  send(topic: string, message: string, options?: amqp.Options.Publish) {
+  //! the second argument in the PUBLISH method NEEDS to be the "key". Before refactor it was the topic. The reason is because this is the routingkey that MUST MATCH the bindingKey
+  //! that is found within the init function
+  async send(topic: string, message: string, options?: amqp.Options.Publish) {
+    console.log('hello from within send!');
+    console.log('this is "this" in the send method', this);
```
Comment (mjhult) [inline packages/library/src/brokers/rabbit/main.ts:184]: Debug code

File: packages/library/src/brokers/rabbit/main.ts [hunk 4]
```diff
@@ -70,40 +95,96 @@ export default class Rabbit extends MessageBroker {
...
+    this.connection = await amqp.connect(this.defaultConfig);
+    if (this.connection === null)
+      throw new Error('No connection for Rabbit. Failed to initialize');
+
+    this.channel = await this.connection!.createChannel();
+    if (this.channel === null)
+      throw new Error('No channel for Rabbit. Failed to initialize');
+
+    // console.log('this is the channel', that.channel);
+    // const that = that;
```
Comment (mjhult) [inline packages/library/src/brokers/rabbit/main.ts:158]: Zombie/Debug

File: packages/library/src/brokers/rabbit/main.ts [hunk 5]
```diff
@@ -43,18 +43,43 @@ export type RabbitListenerOptions = GenericListenerOptions & {
...
     // Let's create a default config, we are also relying on default options for the connect method.
-    const defaultConfig: RabbitClientOptions = {
+
+    // const defaultConfig: RabbitClientOptions = {
+    //   ...connection,
+    //   port: connection.port ?? 5672,
+    //   // protocol: connection.protocol ?? 'amqp',
+    //   // TODO: remove this when we add secure connection support. This overwrites what the user wants.
+    //   protocol: 'amqp',
+    // };
```
Comment (mjhult) [inline packages/library/src/brokers/rabbit/main.ts:80]: Zombie code

**Review by mjhult — CHANGES_REQUESTED** [review]

File: packages/library/src/brokers/rabbit/main.ts (hunk 3 above)
Comment (H3R01A) [reply packages/library/src/brokers/rabbit/main.ts:184]: Removed from the updated code.

File: packages/library/src/brokers/rabbit/main.ts (hunk 4 above)
Comment (H3R01A) [reply packages/library/src/brokers/rabbit/main.ts:158]: Removed from the updated code.

File: packages/library/src/brokers/rabbit/main.ts (hunk 5 above)
Comment (H3R01A) [reply packages/library/src/brokers/rabbit/main.ts:80]: Good catch! Removed in updated code.

File: packages/library/src/brokers/rabbit/main.ts (hunk 2 above)
Comment (H3R01A) [reply packages/library/src/brokers/rabbit/main.ts:53]: Maybe we can talk through this one. Right now the "default" export is "Rabbit"

image

File: packages/library/src/brokers/rabbit/main.ts (hunk 1 above)
Comment (H3R01A) [reply packages/library/src/brokers/rabbit/main.ts]: Removed from the updated code.

**Review by mjhult — APPROVED** [review]

**Pull request merged**